### PR TITLE
tcgc, bug fix on encode string

### PR DESCRIPTION
**Pull request by weidongxu-microsoft**
Problem on TCGC 0.45.1 here https://github.com/Azure/cadl-ranch/pull/656#discussion_r1713119741

link https://github.com/Azure/typespec-azure/issues/1264
link https://github.com/Azure/typespec-azure/issues/1179

### CODE REVIEW

File: packages/typespec-client-generator-core/src/types.ts [hunk 1]
```diff
@@ -215,6 +216,11 @@ export function addEncodeInfo(
       innerType.encode = "bytes";
     }
   }
+  if (isSdkIntKind(innerType.kind)) {
+    if (encodeData && "encode" in innerType) {
+      innerType.encode = getEncodeHelper(context, type, innerType.kind);
```
Comment (tadelesh) [inline packages/typespec-client-generator-core/src/types.ts:221]: i believe `getEncodeHelper` is not used to add encode from typespec, but just a kind of initialization. `addEncodeInfo` is the right place to handle real encode. you may move izzy's change to that function to here.

Comment (weidongxu-microsoft) [reply packages/typespec-client-generator-core/src/types.ts]: Moved the code from `getEncodeHelper` to `addEncodeInfo`.

Added a test case to verify `@encode` on scalar still works.

**Review by tadelesh — APPROVED** [review]

File: packages/typespec-client-generator-core/src/types.ts (hunk 1 above)
Comment (timotheeguerin) [reply packages/typespec-client-generator-core/src/types.ts]: tcgc doesn't support it shouldn't be a reason to not add a scenario. This is a valid scenario which is also the most intuitive way to use `@encode`

Comment (weidongxu-microsoft) [reply packages/typespec-client-generator-core/src/types.ts]: Timothee, not sure I understand your comment.

I've added case of both encode on property
```
      model Test {
        @encode(string)
        value: safeint;
      }
```
and encode on scalar
```
      @encode(string)
      scalar int32EncodedAsString extends int32;

      model Test {
        value: int32EncodedAsString;
      }
```

I assume this is all the cases we had on "integer encode as string'?

Comment (timotheeguerin) [reply packages/typespec-client-generator-core/src/types.ts]: im sorry i commented on the wrong pr after reading the message in the email, forget, does this pr actually fix this?

Comment (timotheeguerin) [reply packages/typespec-client-generator-core/src/types.ts]: yeah was about that one so forget what I said https://github.com/Azure/cadl-ranch/pull/656 seems like you are fixing it

**Review by haolingdong-msft — APPROVED** [review]

**Pull request merged**